### ISSUE TITLE
Test results better than the value reported in paper?

### ISSUE THREAD

**Issue by zechendev** [xincoder/GRIP#1]
hi, thank you for sharing the training code for Apollo.
1. Here is the testing results, It seems Root Mean Square Error (RMSE) for trajectory prediction on NGSIM I-80 and US-101 datasets
#######################################Test
|2020-06-21 17:24:06.688187|[car] All_All: 0.424 0.722 1.050 1.403 1.770 2.179 7.549
|2020-06-21 17:24:06.689078|[human] All_All: 0.152 0.282 0.428 0.588 0.746 0.927 3.122
|2020-06-21 17:24:06.689816|[bike] All_All: 0.419 0.794 1.158 1.568 1.963 2.366 8.269
|2020-06-21 17:24:06.689913|[WS] All_All: 0.265 0.483 0.713 0.967 1.218 1.494 5.140
|2020-06-21 17:24:06.691756|[Test_Epoch49] All_All: 0.264 0.473 0.695 0.952 1.210 1.488 5.080
Is the results correct?

2. Could you please also share the calculation file For ADE and FDE?

**Comment by xincoder** [comment]
> hi, thank you for sharing the training code for Apollo.
> 
> 1. Here is the testing results, It seems Root Mean Square Error (RMSE) for trajectory prediction on NGSIM I-80 and US-101 datasets
>    #######################################Test
>    |2020-06-21 17:24:06.688187|[car] All_All: 0.424 0.722 1.050 1.403 1.770 2.179 7.549
>    |2020-06-21 17:24:06.689078|[human] All_All: 0.152 0.282 0.428 0.588 0.746 0.927 3.122
>    |2020-06-21 17:24:06.689816|[bike] All_All: 0.419 0.794 1.158 1.568 1.963 2.366 8.269
>    |2020-06-21 17:24:06.689913|[WS] All_All: 0.265 0.483 0.713 0.967 1.218 1.494 5.140
>    |2020-06-21 17:24:06.691756|[Test_Epoch49] All_All: 0.264 0.473 0.695 0.952 1.210 1.488 5.080
>    Is the results correct?
> 2. Could you please also share the calculation file For ADE and FDE?

Hi @zechendev, thank you for your interest in our work. 
1. I quickly tested the enhanced idea (GRIP++) on the NGSIM I-80 and US-101 datasets, and simply reported a better result than GRIP (cause I did not have enough time on this project). Then, I focused on the Apollo dataset and tuned the model and training parameters. In this case, you may have already noticed that the reported results on the NGSIM datasets were not generated using this GitHub code. If you directly use this GitHub code on the NGSIM datasets, you may achieve a better result. 
2. We did not evaluate the ADE and FDE by ourselves. All reported results on the Apollo dataset we evaluated by submitting the predicted results to the Apollo website (http://apolloscape.auto/trajectory.html).

**Comment by zechendev** [comment]
Thank you for your reply. After read both code and paper, for the apolloscape dataset, the uploaded code is just B1. If I want to achieve the same results, what should I modified?

**Comment by xincoder** [comment]
> Thank you for your reply. After read both code and paper, for the apolloscape dataset, the uploaded code is just B1. If I want to achieve the same results, what should I modified?

Hi @zechendev, I do not understand the reason why you say "the uploaded code is just B1". You do not need to modify anything. If you follow my instruction to run the training code directly (on a single Nvidia Titan Xp GPU as mine) and then submit the results to the Apollo website, you should get the same results as I mentioned in my ReadMe file. (If you run the training code on a different GPU, you will get different results, but they should be similar to my reported values.)

**Comment by zechendev** [comment]
First, Thank you for you reply. As you mentioned in the paper, the Final WSADE(1.2588) is get by B13(GRIP++) in listed in Table3, the input is velocity. But in the code the input is xy-xy.mean() [it is localization, right?]; They are different. I also have the output for Epoch16. After evaluation using the API, the result is terrible. 
prediction_result.txt is replaced by the grip generated results.
('Load file: ', './test_eval_data/prediction_result.txt')
('Load file: ', './test_eval_data/prediction_gt.txt')
('Load file: ', './test_eval_data/considered_objects.txt')
('WSADE:', 15.452494399775119)
('ADEv, ADEp, ADEb:', 15.751168207181372, 15.35499219314163, 15.438024028712272)
('WSFDE:', 15.707704738103395)
('FDEv, FDEp, FDEb:', 16.205315545826124, 15.426260292875879, 15.997321177591637)

**Comment by xincoder** [comment]
> First, Thank you for you reply. As you mentioned in the paper, the Final WSADE(1.2588) is get by B13(GRIP++) in listed in Table3, the input is velocity. But in the code the input is xy-xy.mean() [it is localization, right?]; They are different. I also have the output for Epoch16. After evaluation using the API, the result is terrible.
> prediction_result.txt is replaced by the grip generated results.
> ('Load file: ', './test_eval_data/prediction_result.txt')
> ('Load file: ', './test_eval_data/prediction_gt.txt')
> ('Load file: ', './test_eval_data/considered_objects.txt')
> ('WSADE:', 15.452494399775119)
> ('ADEv, ADEp, ADEb:', 15.751168207181372, 15.35499219314163, 15.438024028712272)
> ('WSFDE:', 15.707704738103395)
> ('FDEv, FDEp, FDEb:', 16.205315545826124, 15.426260292875879, 15.997321177591637)

@zechendev If you say so, you may not read through and understand the code. The function "preprocess_data" in the main.py file converts all location data into velocity [line 99 in main.py]. Then, I turn the predicted velocity back to location while evaluating [line 206 in main.py]. I do not know how you got the above results, but I did not see any submission has WSADE=15.4 on the Baidu ApolloSape leaderboard (http://apolloscape.auto/leader_board.html -> Trajectory). Thus, I am assuming that you did not submit the results to their website. If you modified my code before, I recommend you download the code and try it from scratch.

**Comment by xincoder** [comment]
@zechendev I uploaded my training log just now so that you can check your outputs. Similar training losses are expected.

**Comment by zechendev** [comment]
Thank you very much. AND SORRY for my misunderstanding for the code.
From the log, both the output are quite similar. It make sense from this point.
|2020-06-21 17:24:03.995962|     Epoch:  49/  50|       Iteration:   67|        Loss:0.71224469|lr: 0.001|
|2020-06-21 17:24:04.044870|     Epoch:  49/  50|       Iteration:   67|        Loss:0.56705058|lr: 0.001|
#######################################Test
|2020-06-21 17:24:06.688187|[car] All_All: 0.424 0.722 1.050 1.403 1.770 2.179 7.549
|2020-06-21 17:24:06.689078|[human] All_All: 0.152 0.282 0.428 0.588 0.746 0.927 3.122
|2020-06-21 17:24:06.689816|[bike] All_All: 0.419 0.794 1.158 1.568 1.963 2.366 8.269
|2020-06-21 17:24:06.689913|[WS] All_All: 0.265 0.483 0.713 0.967 1.218 1.494 5.140
|2020-06-21 17:24:06.691756|[Test_Epoch49] All_All: 0.264 0.473 0.695 0.952 1.210 1.488 5.080

1. What I have done is to uncomment the the test code below. And run main.py
        pretrained_model_path = './trained_models/model_epoch_0016.pt'
        model = my_load_model(model, pretrained_model_path)
        run_test(model, './test_data.pkl')
        to get the prediction_result.txt file.
2. Replace the prediction_result.txt in trajectory_prediction/test_eval_data,
3. run the python evaluation.py --object_file=./test_eval_data/considered_objects.txt --gt_dir=./test_eval_data/prediction_gt.txt --res_file=./test_eval_data/prediction_result.txt
4.paste the results in this Issues.

**Comment by xincoder** [comment]
> Thank you very much. AND SORRY for my misunderstanding for the code.
> From the log, both the output are quite similar. It make sense from this point.
> |2020-06-21 17:24:03.995962| Epoch: 49/ 50| Iteration: 67| Loss:0.71224469|lr: 0.001|
> |2020-06-21 17:24:04.044870| Epoch: 49/ 50| Iteration: 67| Loss:0.56705058|lr: 0.001|
> #######################################Test
> |2020-06-21 17:24:06.688187|[car] All_All: 0.424 0.722 1.050 1.403 1.770 2.179 7.549
> |2020-06-21 17:24:06.689078|[human] All_All: 0.152 0.282 0.428 0.588 0.746 0.927 3.122
> |2020-06-21 17:24:06.689816|[bike] All_All: 0.419 0.794 1.158 1.568 1.963 2.366 8.269
> |2020-06-21 17:24:06.689913|[WS] All_All: 0.265 0.483 0.713 0.967 1.218 1.494 5.140
> |2020-06-21 17:24:06.691756|[Test_Epoch49] All_All: 0.264 0.473 0.695 0.952 1.210 1.488 5.080
> 
> 1. What I have done is to uncomment the the test code below. And run main.py
>    pretrained_model_path = './trained_models/model_epoch_0016.pt'
>    model = my_load_model(model, pretrained_model_path)
>    run_test(model, './test_data.pkl')
>    to get the prediction_result.txt file.
> 2. Replace the prediction_result.txt in trajectory_prediction/test_eval_data,
> 3. run the python evaluation.py --object_file=./test_eval_data/considered_objects.txt --gt_dir=./test_eval_data/prediction_gt.txt --res_file=./test_eval_data/prediction_result.txt
>    4.paste the results in this Issues.

@zechendev In this case, you should check your evaluation code (or GT data) instead of my code. Based on your log info, you should get similar scores as mine. Why don't you submit your results to their website to double-check the evaluation?

**Comment by zechendev** [comment]
Yes, I have upload the prediction_result.txt. And waiting for the result.
Thank you for patience.

**Comment by zechendev** [comment]
![apolloscape_prediction_submission_doublecheck](https://user-images.githubusercontent.com/59218661/85261486-cddf0900-b49e-11ea-8717-db4b45fb0b9d.png)

**Comment by zechendev** [comment]
@xincoder The "./test_eval_data/prediction_gt.txt is just for testing the code which is not the real ground truth. "
I thought I was careless and made a big mistake.
By the way, how long it will take to output the final WSADE?

**Comment by xincoder** [comment]
> @xincoder The "./test_eval_data/prediction_gt.txt is just for testing the code which is not the real ground truth. "
> I thought I was careless and made a big mistake.
> By the way, how long it will take to output the final WSADE?

@zechendev It is the reason why I asked you to submit your results to their website when I first replied to your question. It usually takes 1-5 minutes. However, sometimes their website does not have any response. In this case, I contacted them, and they will check it for you.

**Comment by zechendev** [comment]
@xincoder by email or by creating issues on ApolloScapeAuto/dataset-api?

**Comment by xincoder** [comment]
> @xincoder by email or by creating issues on ApolloScapeAuto/dataset-api?

Via email.

**Comment by zechendev** [comment]
@xincoder Could you please share the email with me? thank you !

**Comment by xincoder** [comment]
> @xincoder Could you please share the email with me? thank you !

Sorry, I do not think it is a good thing to explore others' contact info here. You should be able to find the contact email. I contacted the first Baidu author in their paper.
